## Calculate GWP based on AGWP output

In [1]:
filepath= r"./output/"

In [2]:
import numpy as np
import pandas as pd
#import matplotlib.pyplot as plt
#import matplotlib as mpl
#import itertools
#import matplotlib.patches as mpatches
#from matplotlib import gridspec
pd.set_option('display.float_format', lambda x: '{:,.3f}'.format(x) if abs(x)<0 else ('{:,.2f}'.format(x) if abs(x)<10 else ('{:,.1f}'.format(x) if abs(x)<100 else '{:,.0f}'.format(x))))

In [3]:
#IPCC AR6 values
agwp100_CO2 = 0.0895
agwp20_CO2 = 0.0243 
agwp500_CO2 = 0.314 

In [4]:
model_dict = { 'GFDL_nudge':'GFDL',
               'OSLOCTM3':'OsloCTM',
               'INCA':'INCA',
               'UKCA':'UKCA',
               'WACCM6-2deg':'WACCM',
               'GFDL-emi':'GFDL-emi',
               'OSLOCTM3-emi':'OsloCTM-emi'
             }

sorted_array = ['GFDL','OsloCTM','INCA','UKCA','WACCM','GFDL-emi','OsloCTM-emi']

In [5]:
df_table_agwp = pd.read_csv(filepath+'table_h2_agwp.csv',index_col=0)
df_table_agwp.rename(index=model_dict, inplace=True)
df_table_agwp

h2_agwp_ch4  h2_agwp_o3  h2_agwp_h2o  h2_agwp_ch4ind_o3  \
OsloCTM             0.43        0.22         0.17               0.25   
WACCM               0.45        0.18         0.05               0.28   
INCA                0.43        0.15         0.11               0.14   
GFDL-emi            0.38        0.39         0.25                nan   
OsloCTM-emi         0.44        0.22         0.18               0.26   
UKCA                0.64        0.22          nan               0.24   
GFDL                0.44        0.21         0.16               0.21   

             h2_agwp_ch4ind_h2o  
OsloCTM                    0.06  
WACCM                      0.02  
INCA                       0.04  
GFDL-emi                    nan  
OsloCTM-emi                0.06  
UKCA                        nan  
GFDL                       0.13

In [6]:
rename_columns = {'h2_agwp_ch4':'gwp_ch4',
                  'h2_agwp_o3':'gwp_o3',
                  'h2_agwp_h2o':'gwp_h2o',
                  'h2_agwp_ch4ind_o3':'gwp_ch4ind_o3',
                  'h2_agwp_ch4ind_h2o': 'gwp_ch4ind_h2o'}

## GWP 100

In [7]:
df_table_gwp_100 = df_table_agwp/agwp100_CO2
df_table_gwp_100.rename(columns=rename_columns,inplace=True)
df_table_gwp_100

gwp_ch4  gwp_o3  gwp_h2o  gwp_ch4ind_o3  gwp_ch4ind_h2o
OsloCTM         4.79    2.44     1.94           2.85            0.70
WACCM           5.05    1.97     0.58           3.13            0.28
INCA            4.80    1.64     1.27           1.53            0.46
GFDL-emi        4.26    4.34     2.80            nan             nan
OsloCTM-emi     4.90    2.49     1.97           2.91            0.72
UKCA            7.13    2.46      nan           2.65             nan
GFDL            4.87    2.34     1.80           2.35            1.41

In [8]:
df_table_gwp_100 = df_table_gwp_100.fillna(0.0)
df_table_gwp_100_out = pd.DataFrame(columns=['CH4','O3','StratH2O'],index=df_table_gwp_100.index)

df_table_gwp_100_out['O3'] = df_table_gwp_100['gwp_o3']+df_table_gwp_100['gwp_ch4ind_o3']
df_table_gwp_100_out['CH4']=df_table_gwp_100['gwp_ch4']
df_table_gwp_100_out['StratH2O'] = df_table_gwp_100['gwp_h2o']+df_table_gwp_100['gwp_ch4ind_h2o']
df_table_gwp_100_out=df_table_gwp_100_out.reindex(sorted_array)
df_table_gwp_100_out.replace(0,np.nan,inplace=True)


df_table_gwp_100_out.loc['Model mean'] = df_table_gwp_100_out.drop(['GFDL-emi','OsloCTM-emi']).mean()
df_table_gwp_100_out['Total'] = df_table_gwp_100_out[['O3','CH4','StratH2O']].sum(axis=1) 
df_table_gwp_100_out

CH4   O3  StratH2O  Total
GFDL        4.87 4.69      3.21   12.8
OsloCTM     4.79 5.29      2.64   12.7
INCA        4.80 3.17      1.73   9.70
UKCA        7.13 5.11       nan   12.2
WACCM       5.05 5.10      0.85   11.0
GFDL-emi    4.26 4.34      2.80   11.4
OsloCTM-emi 4.90 5.40      2.69   13.0
Model mean  5.33 4.67      2.11   12.1

Table X: GWP100 for H$_2$ for the individual models. Note that UKCA does not have an estimate for stratospheric H2O. The model means are the mean of the model results for the concentration driven runs, hence excluding OsloCTM-emi and GFDL-emi in the mean. The GWP100 values are calculated by dividing the AGWP100 values with AGWP100 for CO2 of 0.0895 10-12 W m-2 kg-1 yr from IPCC AR6 Table 7.SM.6 

## GWP 500

In [9]:
df_table_gwp_500 = df_table_agwp/agwp500_CO2
df_table_gwp_500.rename(columns=rename_columns,inplace=True)
df_table_gwp_500

gwp_ch4  gwp_o3  gwp_h2o  gwp_ch4ind_o3  gwp_ch4ind_h2o
OsloCTM         1.37    0.70     0.55           0.81            0.20
WACCM           1.44    0.56     0.16           0.89            0.08
INCA            1.37    0.47     0.36           0.44            0.13
GFDL-emi        1.21    1.24     0.80            nan             nan
OsloCTM-emi     1.40    0.71     0.56           0.83            0.20
UKCA            2.03    0.70      nan           0.76             nan
GFDL            1.39    0.67     0.51           0.67            0.40

In [10]:
df_table_gwp_500 = df_table_gwp_500.fillna(0.0)
df_table_gwp_500_out = pd.DataFrame(columns=['CH4','O3','StratH2O'],index=df_table_gwp_500.index)

df_table_gwp_500_out['O3'] = df_table_gwp_500['gwp_o3']+df_table_gwp_500['gwp_ch4ind_o3']
df_table_gwp_500_out['CH4']=df_table_gwp_500['gwp_ch4']
df_table_gwp_500_out['StratH2O'] = df_table_gwp_500['gwp_h2o']+df_table_gwp_500['gwp_ch4ind_h2o']
df_table_gwp_500_out=df_table_gwp_500_out.reindex(sorted_array)
df_table_gwp_500_out.replace(0,np.nan,inplace=True)

df_table_gwp_500_out.loc['Model mean'] = df_table_gwp_500_out.drop(['GFDL-emi','OsloCTM-emi']).mean()

#df_table_gwp_500_out['Total'] = df_table_gwp_500_out['O3'] + df_table_gwp_500_out['CH4'] + df_table_gwp_500_out['StratH2O']
df_table_gwp_500_out['Total'] =  df_table_gwp_500_out[['O3','CH4','StratH2O']].sum(axis=1) 
df_table_gwp_500_out

CH4   O3  StratH2O  Total
GFDL        1.39 1.34      0.91   3.64
OsloCTM     1.37 1.51      0.75   3.62
INCA        1.37 0.90      0.49   2.77
UKCA        2.03 1.46       nan   3.49
WACCM       1.44 1.45      0.24   3.14
GFDL-emi    1.21 1.24      0.80   3.25
OsloCTM-emi 1.40 1.54      0.77   3.70
Model mean  1.52 1.33      0.60   3.45

Table X: GWP500 for H$_2$ for the individual models. Note that UKCA does not have an estimate for stratospheric H2O. The model means are the mean of the model results for the concentration driven runs, hence excluding OsloCTM-emi and GFDL-emi in the mean. The AGWP500 for hydrogen will be identical to the AGWP100 for hydrogen. The GWP500 values are calculated by dividing the AGWP500 with AGWP100 for CO2 of 0.314 10-12 W m-2 kg-1 yr from IPCC AR6 Table 7.SM.6

## GWP 20

For GWP20: 
For AGWP20, perturbations tied to CH4 will still exist at the 10% level after 20 years, and the steady-state approximation can be corrected if we know the time scale of the CH4 perturbation.

The timescale of the perturbation (PT) it the total methane lifetime times the feedback factor (FF). 

Using numbers from Prather et al 2012:

PT = FF * LT = 1.4*9.1

All the RF related to methane decay as exp(-t/PT) where t is time.

Therefore the AGWP_20 = AGWP_100 * (1-exp(20/PT)
for ch4, ch4ind o3 and ch4ind h2o.



In [11]:
PT = 1.47*9.1
print(PT)
adjust_20 =1-np.exp(-20./PT)

print(adjust_20)

13.376999999999999
0.7757746118531375


In [12]:
df_table_agwp_20 = df_table_agwp
df_table_agwp_20['h2_agwp_ch4']=df_table_agwp_20['h2_agwp_ch4']*adjust_20
df_table_agwp_20['h2_agwp_ch4ind_o3']=df_table_agwp_20['h2_agwp_ch4ind_o3']*adjust_20
df_table_agwp_20['h2_agwp_ch4ind_h2o']=df_table_agwp_20['h2_agwp_ch4ind_h2o']*adjust_20

In [13]:
df_table_gwp_20 = df_table_agwp_20/agwp20_CO2
df_table_gwp_20.rename(columns=rename_columns,inplace=True)
df_table_gwp_20

gwp_ch4  gwp_o3  gwp_h2o  gwp_ch4ind_o3  gwp_ch4ind_h2o
OsloCTM         13.7    8.98     7.13           8.14            2.00
WACCM           14.4    7.25     2.12           8.94            0.79
INCA            13.7    6.04     4.69           4.38            1.31
GFDL-emi        12.2    16.0     10.3            nan             nan
OsloCTM-emi     14.0    9.16     7.27           8.32            2.05
UKCA            20.4    9.06      nan           7.59             nan
GFDL            13.9    8.63     6.63           6.71            4.02

In [14]:
df_table_gwp_20 = df_table_gwp_20.fillna(0.0)
df_table_gwp_20_out = pd.DataFrame(columns=['CH4','O3','StratH2O'],index=df_table_gwp_20.index)

df_table_gwp_20_out['O3'] = df_table_gwp_20['gwp_o3']+df_table_gwp_20['gwp_ch4ind_o3']
df_table_gwp_20_out['CH4']=df_table_gwp_20['gwp_ch4']
df_table_gwp_20_out['StratH2O'] = df_table_gwp_20['gwp_h2o']+df_table_gwp_20['gwp_ch4ind_h2o']
df_table_gwp_20_out=df_table_gwp_20_out.reindex(sorted_array)
df_table_gwp_20_out.replace(0,np.nan,inplace=True)

df_table_gwp_20_out.loc['Model mean'] = df_table_gwp_20_out.drop(['GFDL-emi','OsloCTM-emi']).mean()

df_table_gwp_20_out['Total'] = df_table_gwp_20_out[['O3','CH4','StratH2O']].sum(axis=1) 

df_table_gwp_20_out

CH4   O3  StratH2O  Total
GFDL        13.9 15.3      10.6   39.9
OsloCTM     13.7 17.1      9.14   39.9
INCA        13.7 10.4      6.01   30.1
UKCA        20.4 16.6       nan   37.0
WACCM       14.4 16.2      2.91   33.5
GFDL-emi    12.2 16.0      10.3   38.4
OsloCTM-emi 14.0 17.5      9.32   40.8
Model mean  15.2 15.1      7.18   37.5

Table X: GWP20 for H$_2$ for the individual models. Note that UKCA does not have an estimate for stratospheric H2O. The model means are the mean of the model results for the concentration driven runs, hence excluding OsloCTM-emi and GFDL-emi in the mean. For AGWP20, perturbations tied to CH4 will still exist after 20 years, and the steady-state approximation has been corrected. All the RF related to methane decay as exp(-t/PT) where t is time and PT is the timescale of the perturbation, given as the total methane lifetime times the feedback factor (FF). Using numbers from Prather et al 2012 (FF=1.47,LT=9,1) PT is 13.4.
For AGWP20 due to methane and the methane induced ozone and stratospheric water vapor, the AGWP100 values are adjusted by (1-exp(-20/PT)). The AGWP20 values for the rest of the components are as the AGWP100 values. The GWP20 values are calculated by dividing the AGWP20 with AGWP20 for CO2 of 0.0243 10-12 W m-2 kg-1 yr from IPCC AR6 Table 7.SM.6